PBL-30003 & CP-103 app changes not committed

Spacerat opened the pull request:
(Fixes #177)
Cloudpebble's git push command would return "nothing to commit" if the user only added or deleted a resource variant or changed a project setting. 

This PR should fix those issues by adding `has_changed = True` in some appropriate places and by checking for resource files which have been deleted (as was already being done with source files).

Spacerat commented:
Depending on which of #256 and #268 are merged first, the other will need to be changed. Since #256 is Torivon's, I recommend merging that first, then I'll bring this branch up to date.

Spacerat commented:
This is ready to merge now

Katharine commented on ide/tasks/git.py:
```diff
@@ -83,17 +81,23 @@ def github_push(user, commit_message, repo_name, project):
+    expected_paths = set()
+
+    def update_expected_paths(new_path):
+        # This adds the path *and* all parent directories to the list of expected paths.
```
Why?

Spacerat replied:
The directories themselves are in next_tree, so if you don't add them to expected_paths then, when you iterate over next_tree looking for paths which aren't in expected_paths (which represent files to be deleted), it will look as if the directories have been deleted, so has_changed will get set to True.

An alternative to this would be to check (in the "#manage deleted files" block) if the path in next_tree looks like a directory, and if it is, ignore it. However, then you'd need a way to check if a directory actually *was* deleted (e.g. if someone deleted all their workers).

Katharine replied:
Please add this explanation to the code.

Spacerat replied:
Done

Katharine commented:
Given that this breaking would make people particularly upset, what testing have you done on it?

Spacerat commented:
Basically, this: https://github.com/Spacerat/testcommits/commits/master 

I've modified, added and deleted resources, source files, worker files and project settings and pushed the changes. And I've changed nothing and checked that it still says "nothing to commit".
